src/main/java/org/apache/sling/api/wrappers/CompositeValueMap.java

```diff
@@ -18,42 +18,77 @@
  */
 package org.apache.sling.api.wrappers;
 
+import org.apache.sling.api.resource.ValueMap;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.osgi.annotation.versioning.ProviderType;
+
+import java.util.AbstractMap;
 import java.util.Collection;
-import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
-import org.apache.sling.api.resource.ValueMap;
-import org.apache.sling.api.wrappers.impl.ObjectConverter;
-import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
+import static java.util.Arrays.asList;
 
 /**
- * An implementation of the {@link ValueMap} based on two {@link ValueMap}s:
- * - One containing the properties
- * - Another one containing the defaults to use in case the properties map
- *   does not contain the values.
+ * Merge provided {@code ValueMaps} into a single view {@code ValueMap} that aggregates
+ * all key-value pairs of the given maps. The value for a key-value pair is taken from
+ * the first {@code ValueMap} (in iteration order) that has a mapping for the given key.
+ * <br>
+ * Two legacy constructors exist that accept only two {@code ValueMaps}:
+ * <ul>
+ *     <li>One containing the properties</li>
+ *     <li>Another one containing the defaults to use in case the properties map
+ *         does not contain the values.</li>
+ * </ul>
+ * If {@code merge = true}, they work just like when a list of two value maps is passed to the
+ * constructor accepting a list. However, if {@code merge = false} the key-set of the
+ * {@code CompositeValueMap} instance is limited to the keys of the default map.
+ *
  * In case you would like to avoid duplicating properties on multiple resources,
  * you can use a <code>CompositeValueMap</code> to get a concatenated map of
  * properties.
+ * 
  * @since 2.3 (Sling API Bundle 2.5.0)
+ *
+ * @see org.apache.sling.api.resource.ValueMapUtil#merge(List)
+ * @see org.apache.sling.api.resource.ValueMapUtil#merge(ValueMap...)
+ * @see org.apache.sling.api.resource.ValueMapUtil#mergeAndCache(List)
  */
+@ProviderType
```

Comment:
@stefanseifert, @npeltier, @cziegeler I changed CompositeValueMap to be ProviderType. I think it's the right thing to do, but would like to hear if you agree or if you have other opinions.

Reply:
I think its a breaking change

Reply:
then we should probably create a new class parallel to CompositeValueMap. this would also remove the special logic required for the "merge" boolean property which complicates the logic ab it (and i'm not sure for what the merge=false code path is useful, but we have to keep it for backward compatibility).

Reply:
sounds ok to me

Reply:
Leaving it @ConsumerType the addition of a constructor would require a version bump to 3.0.0. So I'll go with the suggestion of @stefanseifert: back the changes out into a separate class (actually an internal class) and allow constructing an instance via the utility's factory method. In the new class I won't support the "merge" property. And finally, I'll mark CompositeValueMap deprecated pointing to the utility instead.

To substitute the "merge" property with value "false", a more generic approach **could** allow filtering the available keys (to be implemented in a separate class/wrapper). But that's for another issue.

I'll also make sure that any exposed classes are @ProviderType.